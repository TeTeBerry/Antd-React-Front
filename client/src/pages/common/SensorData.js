import React from "react";
import axios from "axios";
import FusionCharts from "fusioncharts";
import Charts from "fusioncharts/fusioncharts.charts";
import Widgets from "fusioncharts/fusioncharts.widgets";
import ReactFC from "react-fusioncharts";
import FusionTheme from "fusioncharts/themes/fusioncharts.theme.fusion";
import "bootstrap/dist/css/bootstrap.min.css";
import "./SensorData.css";

ReactFC.fcRoot(FusionCharts, Charts, Widgets, FusionTheme);

class SensorData extends React.Component {
  chartRef = null;
  state = {
    titleData: [],
    showChart: false,
    initValue: 0,
    dataSource: {
      chart: {
        caption: "Real Time Water Flow Rate",
        subCaption: "",
        xAxisName: "Local Time",
        yAxisName: "Flow Rate",
        numberPrefix: "L/min",
        refreshinterval: "2",
        slantLabels: "1",
        numdisplaysets: "10",
        labeldisplay: "rotate",
        showValues: "0",
        showRealTimeValue: "0",
        theme: "fusion"
      },
      categories: [
        {
          category: [
            {
              label: this.clientDateTime().toString()
            }
          ]
        }
      ],
      dataset: [
        {
          data: [
            {
              value: 0
            }
          ]
        }
      ]
    }
  };
  chartConfigs = {
    type: "realtimeline",
    renderAt: "container",
    width: "100%",
    height: "400",
    dataFormat: "json"
  };

  componentDidMount() {
    this.getDataFor();
    this.getTitle();
  }

  getTitle() {
    axios.get("http://127.0.0.1:8088/read.php").then(res => {
      this.setState({
        titleData: res.data
      });
      console.log(res.data[0].reading_time);
    });
  }

  startUpdatingData() {
    this.timerID = setInterval(() => {
      axios.get("http://127.0.0.1:8088/read.php").then(d => {
        let x_axis = this.clientDateTime();
        for (var i = 0; i <= 88; i++) {
          let y_axis = d.data[i].flowRate;
          // if (this.flowRate === undefined) {
          //   return clearInterval(this.timerID);
          // }
          console.log(d.data[i]);
          this.chartRef.feedData("&label=" + x_axis + "&value=" + y_axis);
        }
      });
    }, 3000);
  }

  getDataFor() {
    axios
      .get("http://127.0.0.1:8088/read.php", {
        mode: "cors"
      })
      .then(d => {
        const dataSource = this.state.dataSource;
        dataSource.chart.yAxisMaxValue = 500;
        dataSource.chart.yAxisMinValue = 0;

        dataSource.dataset = d.data;
        console.log(dataSource.dataset);

        this.setState(
          {
            showChart: true,
            dataset: d.data,
            dataSource: dataSource,
            initValue: d.data[0].flowRate
          },
          () => {
            this.startUpdatingData();
          }
        );
      });
  }

  static addLeadingZero(num) {
    return num <= 9 ? "0" + num : num;
  }

  clientDateTime() {
    var date_time = new Date();
    var curr_hour = date_time.getHours();
    var zero_added_curr_hour = SensorData.addLeadingZero(curr_hour);
    var curr_min = date_time.getMinutes();
    var curr_sec = date_time.getSeconds();
    var curr_time = zero_added_curr_hour + ":" + curr_min + ":" + curr_sec;
    return curr_time;
  }

  getChartRef(chart) {
    this.chartRef = chart;
  }

  componentWillUnmount() {
    clearInterval(this.timerID);
  }

  // componentDidMount() {
  //   axios("http://127.0.0.1:8088/read.php").then(response => {
  //     const object = Object.assign(...response.data.map(k => ({ [k]: 0 })));
  //     console.log(object);
  //   });
  // }

  render() {
    return (
      <div className="row mt-5 mt-xs-4">
        <div className="col-12">
          <div className="card custom-card mb-5 mb-xs-4">
            <div className="card-body">
              {this.state.showChart ? (
                <ReactFC
                  {...this.chartConfigs}
                  dataSource={this.state.dataSource}
                  onRender={this.getChartRef.bind(this)}
                />
              ) : null}
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default SensorData;
