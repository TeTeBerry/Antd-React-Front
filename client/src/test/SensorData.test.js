import React from "react";
import setTest from "../setupTests";
import { shallow, render, mount } from "enzyme";
import SensorData from "../pages/common/SensorData";

describe("Test case for setup volume", () => {
  it("should match its empty snapshot", () => {
    const wrapper = shallow(<SensorData />);
    expect(wrapper).toMatchSnapshot();
  });
  it("addLeadingZero", () => {
    const addLeadingZero = jest.fn();
    const wrapper = shallow(<SensorData handleAdd={addLeadingZero} />);
    expect(wrapper.addLeadingZero).toBeUndefined();
  });
  it("clientDateTime", () => {
    const clientDateTime = jest.fn();
    const method = shallow(
      <SensorData handleClientDataTime={clientDateTime} />
    );
    expect(method.clientDateTime).toBeUndefined();
  });
  it("getChartRef", () => {
    const getChartRef = jest.fn();
    const method = shallow(<SensorData getChartRef={getChartRef} />);
    expect(method.getChartRef).toBeUndefined();
  });
  it("getDataFor data", () => {
    const getDataFor = jest.fn();

    const props = {
      getDataFor,
      data: {
        reading_time: "2019-07-03 18:57:35",
        SensorName: "Sensor-1",
        flowRate: "30.23",
        flowMilliLitres: "100.50",
        totalMilliLitres: "23421"
      }
    };

    const wrapper = mount(<SensorData {...props} />);
    expect(wrapper.props().data).toEqual({
      reading_time: "2019-07-03 18:57:35",
      SensorName: "Sensor-1",
      flowRate: "30.23",
      flowMilliLitres: "100.50",
      totalMilliLitres: "23421"
    });
    expect(wrapper.props().getDataFor).toHaveBeenCalledTimes(0);
  });
  it("startUpdatingData", () => {
    const startUpdatingData = jest.fn();
    const wrapper = mount(<SensorData startUpdatingData={startUpdatingData} />);
    expect(wrapper.props().startUpdatingData).toHaveBeenCalledTimes(0);
  });
});
