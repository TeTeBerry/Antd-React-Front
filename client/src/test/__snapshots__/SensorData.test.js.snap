// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test case for setup volume should match its empty snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SensorData />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="col-12 mb-3"
        >
          <div
            className="card-deck custom-card-deck"
          >
            <SensorCard
              alt="fireSpot"
              header="Sensor Name"
              label="(Meter Name)"
              value=""
            />
            <SensorCard
              alt="fireSpot"
              header="Flow Rate"
              label="(L/min)"
              value=""
            />
            <SensorCard
              alt="fireSpot"
              header="Current Liquid Flowing"
              label="(mL/Sec)"
              value=""
            />
            <SensorCard
              alt="fireSpot"
              header="Output Liquid Quantity"
              label="(mL)"
              value=""
            />
          </div>
        </div>,
        <div
          className="col-12"
        >
          <div
            className="card custom-card mb-5 mb-xs-4"
          >
            <div
              className="card-body"
            />
          </div>
        </div>,
      ],
      "className": "row mt-5 mt-xs-4",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="card-deck custom-card-deck"
          >
            <SensorCard
              alt="fireSpot"
              header="Sensor Name"
              label="(Meter Name)"
              value=""
            />
            <SensorCard
              alt="fireSpot"
              header="Flow Rate"
              label="(L/min)"
              value=""
            />
            <SensorCard
              alt="fireSpot"
              header="Current Liquid Flowing"
              label="(mL/Sec)"
              value=""
            />
            <SensorCard
              alt="fireSpot"
              header="Output Liquid Quantity"
              label="(mL)"
              value=""
            />
          </div>,
          "className": "col-12 mb-3",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <SensorCard
                alt="fireSpot"
                header="Sensor Name"
                label="(Meter Name)"
                value=""
              />,
              <SensorCard
                alt="fireSpot"
                header="Flow Rate"
                label="(L/min)"
                value=""
              />,
              <SensorCard
                alt="fireSpot"
                header="Current Liquid Flowing"
                label="(mL/Sec)"
                value=""
              />,
              <SensorCard
                alt="fireSpot"
                header="Output Liquid Quantity"
                label="(mL)"
                value=""
              />,
            ],
            "className": "card-deck custom-card-deck",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "alt": "fireSpot",
                "header": "Sensor Name",
                "label": "(Meter Name)",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "alt": "fireSpot",
                "header": "Flow Rate",
                "label": "(L/min)",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "alt": "fireSpot",
                "header": "Current Liquid Flowing",
                "label": "(mL/Sec)",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "alt": "fireSpot",
                "header": "Output Liquid Quantity",
                "label": "(mL)",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="card custom-card mb-5 mb-xs-4"
          >
            <div
              className="card-body"
            />
          </div>,
          "className": "col-12",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="card-body"
            />,
            "className": "card custom-card mb-5 mb-xs-4",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": null,
              "className": "card-body",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="col-12 mb-3"
          >
            <div
              className="card-deck custom-card-deck"
            >
              <SensorCard
                alt="fireSpot"
                header="Sensor Name"
                label="(Meter Name)"
                value=""
              />
              <SensorCard
                alt="fireSpot"
                header="Flow Rate"
                label="(L/min)"
                value=""
              />
              <SensorCard
                alt="fireSpot"
                header="Current Liquid Flowing"
                label="(mL/Sec)"
                value=""
              />
              <SensorCard
                alt="fireSpot"
                header="Output Liquid Quantity"
                label="(mL)"
                value=""
              />
            </div>
          </div>,
          <div
            className="col-12"
          >
            <div
              className="card custom-card mb-5 mb-xs-4"
            >
              <div
                className="card-body"
              />
            </div>
          </div>,
        ],
        "className": "row mt-5 mt-xs-4",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="card-deck custom-card-deck"
            >
              <SensorCard
                alt="fireSpot"
                header="Sensor Name"
                label="(Meter Name)"
                value=""
              />
              <SensorCard
                alt="fireSpot"
                header="Flow Rate"
                label="(L/min)"
                value=""
              />
              <SensorCard
                alt="fireSpot"
                header="Current Liquid Flowing"
                label="(mL/Sec)"
                value=""
              />
              <SensorCard
                alt="fireSpot"
                header="Output Liquid Quantity"
                label="(mL)"
                value=""
              />
            </div>,
            "className": "col-12 mb-3",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <SensorCard
                  alt="fireSpot"
                  header="Sensor Name"
                  label="(Meter Name)"
                  value=""
                />,
                <SensorCard
                  alt="fireSpot"
                  header="Flow Rate"
                  label="(L/min)"
                  value=""
                />,
                <SensorCard
                  alt="fireSpot"
                  header="Current Liquid Flowing"
                  label="(mL/Sec)"
                  value=""
                />,
                <SensorCard
                  alt="fireSpot"
                  header="Output Liquid Quantity"
                  label="(mL)"
                  value=""
                />,
              ],
              "className": "card-deck custom-card-deck",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "alt": "fireSpot",
                  "header": "Sensor Name",
                  "label": "(Meter Name)",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "alt": "fireSpot",
                  "header": "Flow Rate",
                  "label": "(L/min)",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "alt": "fireSpot",
                  "header": "Current Liquid Flowing",
                  "label": "(mL/Sec)",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "alt": "fireSpot",
                  "header": "Output Liquid Quantity",
                  "label": "(mL)",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="card custom-card mb-5 mb-xs-4"
            >
              <div
                className="card-body"
              />
            </div>,
            "className": "col-12",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div
                className="card-body"
              />,
              "className": "card custom-card mb-5 mb-xs-4",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": null,
                "className": "card-body",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "prevContext": true,
          },
          "getChildContext": Object {
            "calledByRenderer": true,
          },
        },
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
